examples/graph/gat_node_classification.py

```diff
@@ -0,0 +1,402 @@
+            tf.matmul(node_features_expanded, self.kernel_attention)
+        )
+        attention_scores = tf.squeeze(attention_scores, -1)
+
+        # (3) Normalize attention scores
+        attention_scores = tf.math.exp(tf.clip_by_value(attention_scores, -2, 2))
+        attention_scores_sum = tf.math.unsorted_segment_sum(
+            data=attention_scores,
+            segment_ids=edges[:, 0],
+            num_segments=tf.reduce_max(edges[:, 0]) + 1,
+        )
+        attention_scores_sum = tf.repeat(
+            attention_scores_sum, tf.math.bincount(tf.cast(edges[:, 0], "int32"))
+        )
+        attention_scores_norm = attention_scores / attention_scores_sum
+
+        # (4) Gather node states of neighbors, apply attention scores and aggregate
+        node_features_neighbors = tf.gather(node_features_transformed, edges[:, 1])
+        out = tf.math.unsorted_segment_sum(
+            data=node_features_neighbors * attention_scores_norm[:, tf.newaxis],
+            segment_ids=edges[:, 0],
+            num_segments=tf.shape(node_features)[0],
+        )
+        return out
+
+
+class MultiHeadGraphAttention(layers.Layer):
+    def __init__(self, units, num_heads=8, merge_type="concat", **kwargs):
+        super().__init__(**kwargs)
+        self.num_heads = num_heads
+        self.merge_type = merge_type
+        self.attention_layers = [GraphAttention(units) for _ in range(num_heads)]
+
+    def call(self, inputs):
+        atom_features, pair_indices = inputs
+
+        # Obtain outputs from each attention head
+        outputs = [
+            attention_layer([atom_features, pair_indices])
+            for attention_layer in self.attention_layers
+        ]
+        # Concatenate or average the node states from each head
+        if self.merge_type == "concat":
+            outputs = tf.concat(outputs, axis=-1)
+        else:
+            outputs = tf.reduce_mean(tf.stack(outputs, axis=-1), axis=-1)
+        # Activate and return node states
+        return tf.nn.relu(outputs)
+
+
+def GraphAttentionNetwork(hidden_units, num_heads, num_layers):
+
+    node_features = layers.Input((NODE_DIM), dtype="float32")
+    adjacency = layers.Input((2), dtype="int32")
+
+    x = layers.Dense(hidden_units * num_heads, activation="relu")(node_features)
+
+    for _ in range(num_layers):
+        x_updated = MultiHeadGraphAttention(hidden_units, num_heads)([x, adjacency])
+        x = layers.Add()([x_updated, x])
+
+    out = layers.Dense(OUTPUT_DIM)(x)
+
+    model = keras.Model(inputs=[node_features, adjacency], outputs=[out])
+
+    return model
+
+
+# Build graph attention network model
+gat_model = GraphAttentionNetwork(HIDDEN_UNITS, NUM_HEADS, NUM_LAYERS)
+
+# Visualize built model
+keras.utils.plot_model(gat_model, show_shapes=True, show_layer_names=False)
+
+"""
+### Define custom `train_step`
+"""
+
+
+def get_train_step_fn(model, loss_fn, optimizer):
+    @tf.function
+    def train_step(inputs, labels, indices):
```

Comment:
You should be able to structure this as a `train_step` method on a subclassed `Model`, I would think. It would make the code more reusable.

You could just:

```
class GraphAttentionNetwork(keras.Model):

    def train_step(self, data):
        inputs, labels, indices = data
        ...

...
model = GraphAttentionNetwork(inputs=[node_features, adjacency], outputs=[out])
```

ref: https://keras.io/guides/customizing_what_happens_in_fit/#going-lowerlevel

Reply:
Sorry for my ignorance: It is not known to me how you pass three inputs (a three-tuple) to`model.fit()`. Normally the `train_step` would look like this: 
```
def train_step(self, data):
   x, y = data
```
or 
```
def train_step(self, data):
   x, y, sample_weight = data
```
However, sample weight is passed to the argument `sample_weight` in the `.fit()` method.

I also stumble upon this problem (as `node_features` and `adjacency` are of different size/shape:
```
ValueError: Data cardinality is ambiguous:
  x sizes: 2708, 5429,
  ```

Reply:
I have a solution to this (make the code more reusable; see commit in near future). Though it has to be somewhat different from a typical implementation as we're working with this one single graph.

Reply:
If you cannot iterate on the data in pairs (e.g. inputs, labels, indices having the same number of entries) then I agree that a low-level training loop is preferable.